docs/adr/2022-04-26__alerting.md

```diff
@@ -0,0 +1,295 @@
+
+Control limits are determined by calculating the standard deviation:
+
+```math
+stdev(v)_t = \sqrt(EWMA(v^2)_t - EWMA(v)^2_t)
+```
+
+If the recorded values for a metric exceeds `EWMA_t + stdev(v)_t` or is below `EWMA_t - stdev(v)_t`, then a change is present.
+
+This method is memory and CPU efficient and can be easily implemented in SQL. It only works for metrics with no seasonality. In the case of OpMon most operational metrics, like memory consumption or different load times, are not influenced by different week days or holidays.
+
+A few drawbacks of this methods are that over time the control limits can get temporarily inflated after a spike and changes could go unnoticed afterwards. Slow changes will go unnoticed, for example if a metric slowly degrades over a long period of time, the average will adjust to it instead of noticing the change.
+Another issue is that for computing the percentiles for each metric, values need to get bucketed. Depending on the resolution of these buckets, value changes do not show up as steady changes but instead as sudden jumps. This could potentially trigger this check if the resolution of the bucketes is low.
+
+This check needs to be configured for metrics like:
+
+```toml
+[alert]
+
+[alert.crash_changes] # new alert
+type = "ema"
+precentiles = [50, 90] # percentiles that should be considered
+weight = 0.4    # weight factor    
+probes = [      # list of metrics to check
+    "startup_crashes", 
+    "main_crashes", 
+    "oom_crashes"
+]
+```
+
+
+## Displaying Changes and Notifications
+
+Significant changes are displayed in a table on the OpMon dashboard. For example:
+
+<img src="./images/alerting_table.png" width="500"/>
+
+Users can opt-in to receiving notifications via email by creating a new alert on the table in Looker:
+
+<img src="./images/alerting_notification.png" width="500"/>
+
+
+
+## Alternatives Considered
+
+### More Advanced Time Series Modelling
+
+Instead of using EWMA there are more advanced models, such as ARIMA models, which can be used to express more complicated characteristicts of time series. The idea is to fit or train these models to sample data and then use them to make predictions or draw conclusions on whether a significant change of certain metrics has occured. 
+
+BigQuery has [support for creating ARIMA models](https://cloud.google.com/bigquery-ml/docs/reference/standard-sql/bigqueryml-syntax-create-time-series), however the price for using BigQuery ML is quite significant ($250 per TB of processed data). A separate model would be needed for every OpMon project and every metric that uses this type of alerting which would either result in significant cost or computation time (is we, for example, use an external Python library instead).
+
+### Comparing Windows of Averages
```

Comment:
Personally, I think the 7-day moving average idea has a few advantages. It's more interpretable and easier to understand the definitions of (and thus correctly calculate) the control limits. I don't _think_ the exponentially weighted moving average method will avoid the alert spike issue mentioned below, so both will suffer from that.

Reply:
The prototype that I implemented for a couple of existing opmon projects was less noisy when using EWMA. However, we can also start simple with the 7-day moving average and implement the more complex EWMA mechanism later if it turns out that the checks are too noisy.